### PR TITLE
Default to msgpack for safetensors

### LINKED ISSUE #27416: 4.35.0 requires troublesome pytorch install for JAX users
### System Info

- `transformers` version: 4.35.0
- Platform: Linux-5.19.0-45-generic-x86_64-with-glibc2.35
- Python version: 3.11.6
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): not installed (NA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): 0.7.4 (gpu)
- Jax version: 0.4.20
- JaxLib version: 0.4.20
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help?

@sanchit-gandhi

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

`FlaxT5ForConditionalGeneration.from_pretrained("t5-base")`

### Expected behavior

I was using transformers 4.31.0 and loading pre-trained models worked pretty smoothly. I recently tried to upgrade to 4.35.0 ad encountered a cluster of issues:

- The same `from_pretrained` call fails at runtime now and says that pytorch must be installed (I assume this is due to the `safetensors` changes mentioned here: https://github.com/huggingface/transformers/releases/tag/v4.35.0)
- I don't see any way to disable the new loading behavior and revert to the old loading behavior
- I install pytorch (sort of unfortunate that a third-party lib is mandating the install of an entire separate DL stack)
- However, now `jax.devices()` warns about a failure to initialize CUDA and only returns the CPU as a device. This is because `pytorch` pins specific versions of CUDA support libs (https://github.com/pytorch/pytorch/blob/ceb07656c234b78bd045312f871b9cdca8c3c3ba/.github/scripts/generate_binary_build_matrix.py#L32) which are incompatible with those JAX wants.
- I also tried some other permutations of pytorch and JAX versions (between 0.4.14 and 0.4.20 IIRC). None of them are compatible. They all fail at runtime with either JAX's failure to initialize CUDA or throw an error about JAX being compiled against newer versions of the libs.

TL;DR: 4.35.0 requires `torch` alongside `jax` to load models and it's hard to do this in a way that satisfies both sets of constraints for CUDA support libs.

This was ultimately resolvable for me by installing the CPU version of pytorch, but I thought I'd write out this issue for anyone else that encounters it and so that perhaps this interaction can be handled and/or documented more fluently.

### PR DESCRIPTION
**Pull request by LysandreJik**
Fix https://github.com/huggingface/transformers/issues/27416

### CODE REVIEW

File: tests/test_modeling_tf_utils.py [hunk 1]
```diff
@@ -535,6 +535,71 @@ def test_safetensors_tf_from_sharded_h5_with_sharded_safetensors_hub(self):
...
+
+        # Can load from the PyTorch-formatted checkpoint
+        safetensors_model = TFBertModel.from_pretrained("hf-internal-testing/tiny-bert-pt-safetensors")
+        for p1, p2 in zip(tf_model.weights, safetensors_model.weights):
+            self.assertTrue(np.allclose(p1.numpy(), p2.numpy()))
+
+    @require_safetensors
+    def test_safetensors_load_from_local_from_safetensors_pt(self):
+        """
+        This test checks that we can load safetensors from a checkpoint that only has those on the Hub.
```
Comment (amyeroberts) [inline tests/test_modeling_tf_utils.py:570]:
```suggestion
        This test checks that we can load safetensors from a local checkpoint that only has those
```

File: tests/test_modeling_tf_utils.py [hunk 2]
```diff
@@ -535,6 +535,71 @@ def test_safetensors_tf_from_sharded_h5_with_sharded_safetensors_hub(self):
...
+        """
+        This test checks that we'll first download msgpack weights before safetensors
+        The safetensors file on that repo is a pt safetensors and therefore cannot be loaded without PyTorch
+        """
+        TFBertModel.from_pretrained("hf-internal-testing/tiny-bert-pt-safetensors-msgpack")
+
+    @require_safetensors
+    def test_safetensors_load_from_local_msgpack_before_safetensors(self):
+        """
+        This test checks that we'll first download msgpack weights before safetensors
```
Comment (amyeroberts) [inline tests/test_modeling_tf_utils.py:596]:
```suggestion
    def test_safetensors_load_from_local_h5_before_safetensors(self):
        """
        This test checks that we'll first download h5 weights before safetensors
```

File: tests/test_modeling_tf_utils.py [hunk 3]
```diff
@@ -535,6 +535,71 @@ def test_safetensors_tf_from_sharded_h5_with_sharded_safetensors_hub(self):
...
+            location = snapshot_download("hf-internal-testing/tiny-bert-pt-safetensors", cache_dir=tmp)
+            safetensors_model = TFBertModel.from_pretrained(location)
+
+        for p1, p2 in zip(tf_model.weights, safetensors_model.weights):
+            self.assertTrue(np.allclose(p1.numpy(), p2.numpy()))
+
+    @require_safetensors
+    def test_safetensors_load_from_hub_msgpack_before_safetensors(self):
+        """
+        This test checks that we'll first download msgpack weights before safetensors
```
Comment (amyeroberts) [inline tests/test_modeling_tf_utils.py:588]:
```suggestion
    def test_safetensors_load_from_hub_h5_before_safetensors(self):
        """
        This test checks that we'll first download h5 weights before safetensors
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

**Comment by LysandreJik** [comment]
Nice catches!

**Pull request merged**